Fix MentionBlot display

lunaru opened the pull request:
Change `innerText` to `innerHTML` to avoid flattening the `denotationChar` span. This retains the className for ql-mention-denotation-char so that it's style-able

lunaru committed:
Fix MentionBlot display

Change `innerText` to `innerHTML` to avoid flattening the `denotationChar` span. This retains the className for ql-mention-denotation-char so that it's style-able

lunaru commented:
@MadSpindel would appreciate a quick review and incorporation of this very small fix that makes the denotation char much easier to style via CSS and seems to maintain the original intent of the code

lunaru commented:
@MadSpindel wanted to follow up here to see if we could get this fix merged in

lunaru commented:
@MadSpindel yes, extending MentionBlot with a custom Blot is what we're doing in our codebase right now to work around this.

However, I think this particular PR fix is still the right thing to do because of 2 reasons:

1. If you look at the code history, this was `innerHTML` before and was changed in a cleanup/refactor, which I believe was unintentional
2. The surrounding code clearly sets html text and then the `innerText` removes that HTML, signifying that the current behavior is not actually intentional. i.e. lines 53 to 56 are undone by `innerText`

MadSpindel commented:
It was changed in PR #341. Not sure why you want to change this if you are already using custom blot?

lunaru commented:
I put up this PR because of the reasons stated. The code doesn't seem to be doing what it should be intuitively doing and it's undoing its own span creation. The `innerText` wipes out the `span` created in lines 53-56. 

Would there be appetite for a different spin of this fix if we continue to use innerText for the XSS concern but wrapped in a separate `span` so that it doesn't nuke the denotation span?

MadSpindel commented:
Yes, that sounds like a good solution.

lunaru committed:
Use separate span for mention value

Avoids flattening the blot HTML with an innerText over the entire element and instead inserts via innerText on an internal span.

lunaru commented:
@MadSpindel okay PR updated, I think this keeps the XSS concern at bay while also making sure the blot keeps its internal HTML structure as originally intended

MadSpindel commented:
@csculley @thexeos Any opinions?

oliviasculley commented:
Hello, I was previously using the `csculley` account but I've switched to using this one for everything.

This PR will fix #357, basically when we switched to `innerText` to prevent XSS in most people's usecases, we also ended up removing an extra span element in the mention that might've broken people's styling, but luckily that was done on a major version release so hopefully that was expected.

Since we're adding an extra node instead of removing one, I think it might not be a big issue to release it under a new minor version rather than a major version, but it could break anyone who has styling changes that depend on the structure rather than the class names, but I don't think that will be a major bug in most cases

lunaru commented:
Thanks all!
